Source/WTF/wtf/ThreadSafeWeakPtr.h

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright (C) 2022 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <wtf/Lock.h>
+#include <wtf/MainThread.h>
+#include <wtf/RefPtr.h>
+
+namespace WTF {
+
+template<typename> class ThreadSafeWeakPtr;
+template<typename, DestructionThread> class ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr;
+
+template<typename T>
+class ThreadSafeWeakPtrControlBlock {
+    WTF_MAKE_NONCOPYABLE(ThreadSafeWeakPtrControlBlock);
+    WTF_MAKE_STRUCT_FAST_ALLOCATED;
+public:
+    void ref() const
+    {
+        Locker locker { m_lock };
+        m_weakReferenceCount++;
+    }
+
+    void deref() const
+    {
+        bool shouldDeleteControlBlock { false };
+        {
+            Locker locker { m_lock };
+            RELEASE_ASSERT_WITH_SECURITY_IMPLICATION(m_weakReferenceCount);
+            if (!--m_weakReferenceCount && !m_strongReferenceCount)
+                shouldDeleteControlBlock = true;
+        }
+        if (shouldDeleteControlBlock)
+            delete this;
+    }
+
+    void strongRef() const
+    {
+        Locker locker { m_lock };
+        RELEASE_ASSERT_WITH_SECURITY_IMPLICATION(m_object);
```

Comment:
Is it possible to avoid locking here, if m_strongReferenceCount is atomic?

I guess not, due to the clang threadsafety annotations, and it's likely preferable to have those to prevent bugs, rather than the slight performance win.

Reply:
I'm pretty sure we would still need to lock to handle multiple threads creating and destroying weak and strong pointers and converting between the two.